tsdb/head.go

```diff
@@ -1930,11 +1917,71 @@ func (s *stripeSeries) gc(mint int64, minOOOMmapRef chunks.ChunkDiskMapperRef) (
 		deleted[storage.SeriesRef(series.ref)] = struct{}{}
 		series.lset.Range(func(l labels.Label) { affected[l] = struct{}{} })
 		s.hashes[hashShard].del(hash, series.ref)
-		delete(s.series[refShard], series.ref)
 		deletedForCallback[series.ref] = series.lset
+
+		delete(s.series[refShard], series.ref)
+	}
+
+	s.iterSripesForGC(checkSeriesFromStripe)
+
+	// For survivors series, truncate old chunks and check if any chunks left.
+	// If survivors, move it to survivors.
+	// If not, we need to remove it from the hashes.
+	deletedSurvivorsForCallback := make(map[chunks.HeadSeriesRef]labels.Labels)
+
+	// Check survivors.
+	for _, series := range prevSurvivors {
+		series.Lock()
+
+		rmChunks += series.truncateChunksBefore(mint, minOOOMmapRef)
+		var keep bool
+		actualMint, minOOOTime, minMmapFile, keep = shouldKeepMemSeries(series, actualMint, minOOOTime, minMmapFile)
+		if keep {
+			survivors[series.ref] = series
+			series.Unlock()
+			continue
+		}
+
+		hash := series.lset.Hash()
+		hashShard := hash & uint64(s.size-1)
+
+		s.locks[hashShard].Lock()
+		deleted[storage.SeriesRef(series.ref)] = struct{}{}
+		series.lset.Range(func(l labels.Label) { affected[l] = struct{}{} })
+		s.hashes[hashShard].del(hash, series.ref)
+		deletedSurvivorsForCallback[series.ref] = series.lset
+		s.locks[hashShard].Unlock()
+
+		series.Unlock()
+	}
+	// Update survivors.
+	s.survivors.Store(survivors)
```

Comment:
Can we keep track of the size of this map plus the average size of the `series` map that is reseted below in line 1970?

I think that can help a user set the stripe size as well as knowing more about whats happening during gc.

Reply:
Done. I added a tracker for stripe series, we can deduce the rest of the numbers from that (suvivors are total - stripe series)